### PR TITLE
[WIP] layers: list: organize languages

### PR DESCRIPTION
**Pull request by Anton-Latukha**
List of language layers in Spacemacs is huge. Pull request organizes languages.

Reason to do so came from personal experience of configuration. List was long enumeration of everything and their uncle. As DevOps needs to look in many types of files/data, I needed to look several times through it to be sure I picked all that is helpful to me.

Also there is a factor of: "Interesting, what that language is for" and "I forgot what that language is for".

#### Categorization
Yes, there are a huge number of ways about how to categorize programming languages.

I thought about classifications of this list, and from time to time studied classification of languages, for a half of the year.

And this is the conclusion.

#### General-purpose OR Domain-specific
There are two major defined domains of computer languages:
* General-purpose programming languages (system programming, general-purpose scripting languages included)
* Domain-specific languages

#### Paradigm classification
All _programming_ languages by themselves have solid [paradigm classification](https://en.wikipedia.org/wiki/Comparison_of_programming_paradigms).

#### Additional points
Language paradigm choices are targeted to fulfill the areas of application of the language. So areas of application that language created for has an influence on design of language and paradigms chosen. That is why area of application is primary to the paradigm.

Language can be _purely functional_, and have very narrow domain-specific application (Coq, Faust), making it not possible/useful in other areas.

Many of _domain-specific_ languages are not _Turing-complete_ (_programming_ in the strict sense) languages (example - markup languages). That is other point why area (target) of application is primary to the paradigm.

Many languages are multi-paradigm, due to they are had been multi-paradigm from the start design, or because of the imperative -> functional paradigm creep, or due language having not purely functional design, that enables it to have other paradigms.

To move language from multi-paradigm to imperative or other group we need to have a field professional statement, that the language possibilities in some paradigms are very rudimentary, and that language is primary belongs to one paradigm. Otherwise language belongs to multi-paradigm.

Many general-purpose languages initially were/are scripting languages, but most of general-purpose scripting languages are _Turing-complete_ and offer general-purpose grade programming libraries/tooling, and also now have compilation possibilities, and other traits of classical general-purpose languages. That is why general-purpose scripting languages group is not formed, because it is hard to strictly state they are _scripting_ - they are general-purpose languages.

Major C family group is also hard to form. It is OR half of languages in broad sense, OR (`c-c++` & `d`) in strictest sense, that makes a group too small (only two layers).  

Where strict sub-goups of more then two by the main trait could be created - they was grouped (JavaScript (since purpose of TypeScript and PureScript is to produce good JavaScript output), Lisp dialects (and link to Emacs Lisp being DSL), DSL -> {Markup, {Windows, Unix} Scripring}).

Also there are a couple of explanations, so mortal-ones can instantly see that `ess` is a bundle for `R`, and what is in `gpu` layer.

I am open to make additional touches in the name of the strict sense.

### CODE REVIEW

**Comment by syl20bnr** [comment]
Great work! I would like to see this in Spacemacs but it needs more work on the Spacemacs side because the `layers.org` file is generated automatically by a script from the file tree!
 
So this great work will be overwritten when the file is regenerated :-(

I'm not fond of moving all the files physically to get this classification but I like the fact that the `layers.org` file is generated, so we need a logical solution to be able to do both.

A first solution that comes to mind is to define a new conventional variable `xxx-sub-categories` which takes a list of sub-categories from root to leaf:

```
(setq clojure-sub-categories '("Multi-paradigm" "List dialects"))
```

File: layers/LAYERS.org [hunk 1]
```diff
@@ -67,64 +77,81 @@
 - [[file:+intl/chinese/README.org][chinese]]
 - [[file:+intl/keyboard-layout/README.org][keyboard-layout]]
 
-* Programming and markup languages
-- [[file:+lang/agda/README.org][agda]]
-- [[file:+lang/asciidoc/README.org][asciidoc]]
-- [[file:+lang/asm/README.org][asm]]
+* General-purpose and domain-specific languages
```
Comment (syl20bnr) [inline layers/LAYERS.org]: Let's call it simply `Programming Languages` and have a `Markup Languages` at the same level.

File: layers/LAYERS.org [hunk 2]
```diff
@@ -67,64 +77,81 @@
...
+- [[file:+lang/purescript/README.org][purescript]]
 - [[file:+lang/typescript/README.org][typescript]]
-- [[file:+lang/vimscript/README.org][vimscript]]
-- [[file:+lang/windows-scripts/README.org][windows-scripts]]
-- [[file:+lang/yaml/README.org][yaml]]
+
+**** Lisp dialects
+- [[file:+lang/clojure/README.org][clojure]]
+- [[file:+lang/common-lisp/README.org][common-lisp]]
+- Emacs Lisp: look [[Domail-specific languages]]
```
Comment (syl20bnr) [inline layers/LAYERS.org]: Let's find a solution without cross-references. I guess that `Lisp dialects` is the most important category for `Emacs Lisp` so we can put it in this category only. Moreover we can say that any lisp is also a DSL as any lisp can be turned into a DSL easily.

**Comment by syl20bnr** [comment]
Instead of `xxx-sub-categories` let's call it `xxx-categories` and leave the physical organization of the layer completely independent from it.

So it becomes: 

```
(setq clojure-categories '("Programming Languages" "Multi-paradigm" "List dialects"))
```

Maybe if the event where there is no `xxx-categories` for a layer we can fallback to the physical organization.

File: layers/LAYERS.org (hunk 2 above)
Comment (Anton-Latukha) [reply layers/LAYERS.org]: Done.

File: layers/LAYERS.org (hunk 1 above)
Comment (Anton-Latukha) [reply layers/LAYERS.org]: Done. `Markup languages` now top-level entry as `Programming Languages`.

**Comment by Anton-Latukha** [comment]
1. In `Programming languages` `Markup languages` sub-entries reduced the words in names (aka removed ...` languages`.
  That is what probably meant `Instead of 'xxx-sub-categories' let's call it 'xxx-categories'`.

2. Added a optional commit that has alphabetical org sort check, all it did holds - swapped `Emacs` and `E-mail`.

**Comment by syl20bnr** [comment]
@JAremko is working on improving the categorization of layers and auto-generation of layers list. We may add tags via the org file of the layer.

**Comment by Anton-Latukha** [comment]
@JAremko 

Good day.

This PR in brief - categorization of program language layers by language families and paradigms.

Above comment reference your work on layers lists infrastructure.

I have an entry ELisp skill, and knowledge in functional programming.
Please, if you can direct me with meaningful lower-level instructions (AKA in terms of lists and functions), how layers list generation works now, then I would be able to make this change.

Thank you.

**Comment by JAremko** [comment]
@Anton-Latukha  Most of documentation processing written in Clojure and resides in the separate repo: https://github.com/JAremko/spacetools 
l'm planning to rewrite legacy LAYERS.org generation this month then we'll have a way to make the structure that you created persistent. Currently it is second in my TODO list :wink:

**Comment by Anton-Latukha** [comment]
Ok, thank you. Then my stall wasn't in vein =)

If you would deside on the fly - feel free to implement this.

If you would leave this for me - I would ping you in 2 month for general directions.

Thank you.

**Comment by JAremko** [comment]
@Anton-Latukha I added new layer generation system (it's pretty much WIP)

It is driven by tags of `README.org` files:
``` org-mode
#+TITLE: Forth layer

#+TAGS: general|imperative|lang|layer

* Table of Contents                     :TOC_4_gh:noexport:
- [[#description][Description]]
  - [[#features][Features:]]
- [[#install][Install]]
- [[#key-bindings][Key bindings]]

* Description
This layer adds basic support for the Forth family of languages to spacemacs.
...
```

You can try generating layers file locally with:
``` sh
docker run --rm \
  -v <SPACEMACS_REPO_ROOT>:/tmp/docs \
  -v <PATH_TO_CONFIG_FILE>:/opt/spacetools/spacedoc-cfg.edn \
  jare/spacetools docfmt /tmp/docs/
```

Config file looks like this:
``` clojure
{:spacetools.spacedoc.config/layers-org-query
 [{"lang" [
           "pure" {"general" [
                              "imperative"
                              "multi-paradigm"
                              ]
                   }
           ]
   }
  "markup"
  ]

 :spacetools.spacedoc.config/valid-tags
 {"lang" "Programming languages"
  "js" "JavaScript"
  "dsl" "Domain-specific"
  "e-mail" "E-mail"
  "distribution" "Distributions"
  "os" "Operating systems"
  "chat" "Chats"
  "misc" "Misc"
  "framework" "Frameworks"
  "markup" "Markup languages"
  "tagging" "Tags"
  "i18n" "internationalization"
  "pairing" "Pair programming"
  "web service" "Web services"
  "versioning" "Source control"
  "fun" "Fun"
  "script" "Scripting"
  "imperative" "Imperative"
  "general" "General-purpose"
  "pure" "Purely functional"
  "emacs" "Emacs"
  "vim" "Vim"
  "completion" "Completion"
  "tool" "Tools"
  "extra" "Extra"
  "lisp" "Lisp dialects"
  "layer" "layers"
  "theme" "Themes"
  "multi-paradigm" "Multi-paradigm"
  "checker" "Checkers"
  }}
```

`:spacetools.spacedoc.config/valid-tags` specifies what tags are allowed (keys of the map) and what names(values of the map) they'll have in `LAYERS_WIP.org` file.

`:spacetools.spacedoc.config/layers-org-query` is a sort of a database like query:

Here maps `{<KEY> <VAL>}` specify a group of layers/groups (will become a headline in `LAYERS_WIP.org` file). Vectors `[ A B C D ... ]` populate the group. Vector can contain another map (group) or a string.  If it's a string then all layers satisfying full path query will be pooled in its place.

For example query like this `["lang"]` will simply pull all layers that have `lang` tag.

- - - - - - - - - -

Important notes:
- You can query for the same REDME.org files as many times as you want. So you can provide alternative groupings.
- Tags are case sensitive.
- Tags can have spaces but not `|` character (the're used to separate tags in `README.org` files)
- String values in `:spacetools.spacedoc.config/layers-org-query` are tags and must be present in ` :spacetools.spacedoc.config/valid-tags`
- Empty `LAYERS_WIP.org` will cause an error. It will be empty if no query is satisfied.
- `layer` tag is implicit in `:spacetools.spacedoc.config/layers-org-query`  as its root. So only README.org files that have layer tag will be considered.
-  Not sure if it's buggy but I'll be back in a couple of days and hopefully will have time next week to fix them. :bug:   :hammer:

**Comment by JAremko** [comment]
@Anton-Latukha  feel free to edit https://github.com/syl20bnr/spacemacs/blob/develop/.ci/spacedoc-cfg.edn and add tags to `README.org` files.  Bot will update https://github.com/syl20bnr/spacemacs/blob/develop/layers/LAYERS_WIP.org automatically.

**Comment by JAremko** [comment]
Also [this](https://github.com/syl20bnr/spacemacs/blob/develop/.ci/spacedoc-cfg.edn#L3):
``` clojure
 [{"lang" [
           "pure" {"general" [
                              "imperative"
                              "multi-paradigm"
                              ]
                   }
           ]
   }
  "markup"
  ]
```

should be formatted like this:

``` clojure
[{"lang" [
          "pure"
          {"general" [
                      "imperative"
                      "multi-paradigm"
                      ]
           }
          ]
  }
 "markup"
 ]
```
I missformatted it.

**Comment by JAremko** [comment]
@Anton-Latukha  I added almost all layers #12254 . If you want, you can take care of languages  :wink:

**Comment by Anton-Latukha** [comment]
Ok.
Seems understandable.
I annotating your explanation and can say that currently trying to make changes.
I would work today and tomorrow, and bring changes or report.

**Comment by JAremko** [comment]
@Anton-Latukha  This evening I'll add "layer" and "uncategorized" tags to all layers that need to be tagged. So you'll just have to remove "uncategorized" and add proper tags to the README.org files.

**Comment by Anton-Latukha** [comment]
@JAremko 
Understood and observed everything I needed.
And started working.

Expliit `layer` is a good idea.
Would continue after you merge `uncategorized`.

**Comment by JAremko** [comment]
@Anton-Latukha  rebese your fork https://github.com/syl20bnr/spacemacs/pull/12259

**Comment by JAremko** [comment]
@Anton-Latukha  all that still needs a tag is under https://github.com/syl20bnr/spacemacs/blob/develop/layers/LAYERS_WIP.org#readmeorg-files-that-need-proper-tags

**Comment by JAremko** [comment]
@Anton-Latukha  added new tags to the query and valid tags https://github.com/syl20bnr/spacemacs/pull/12260/files

**Comment by Anton-Latukha** [comment]
Currently, I pushed the 100% copy of structure that was published/discussed here previously, now in the new format.

I am still thinking, maybe to include Markup languages as `Programming languages/DSL/Markup languages`, as theoretically they are such (aka computer languages).

**Comment by Anton-Latukha** [comment]
I am still thinking, maybe to include Markup languages as `Programming languages/DSL/Markup languages`, as theoretically they are such (aka computer languages).

**Comment by Anton-Latukha** [comment]
Layers that were introduced in the mean time are in the process.

**Comment by Anton-Latukha** [comment]
As we see - central CI is ok with changes.
Suggested command for generating layers file locally with: 
```fish
docker run --rm \
  -v /home/pyro/.emacs.d:/tmp/docs \
  -v /home/pyro/.emacs.d/.ci/spacedoc-cfg.edn:/opt/spacetools/spacedoc-cfg.edn \
  jare/spacetools docfmt /tmp/docs/
```
Yiellded partial result:
<details><summary>Full report</summary>
<p>

```
======================================================================
Exporting documentation for formatting with SDN tools:
======================================================================
PROCESS: #<process worker<6>>

Exporting "/tmp/docs/doc/DOCUMENTATION.org" into "/tmp/spacetools/doc/DOCUMENTATION.sdn"
Successfully exported "/tmp/docs/doc/DOCUMENTATION.org"
Process worker<6> has finished.

PROCESS: #<process worker<7>>

Exporting "/tmp/docs/CHANGELOG.org" into "/tmp/spacetools/CHANGELOG.sdn"
Successfully exported "/tmp/docs/CHANGELOG.org"
Process worker<7> has finished.

PROCESS: #<process worker<5>>

Exporting "/tmp/docs/layers/+tools/node/README.org" into "/tmp/spacetools/layers/+tools/node/README.sdn"
Successfully exported "/tmp/docs/layers/+tools/node/README.org"
Exporting "/tmp/docs/layers/+spacemacs/spacemacs-editing-visual/README.org" into "/tmp/spacetools/layers/+spacemacs/spacemacs-editing-visual/README.sdn"
Successfully exported "/tmp/docs/layers/+spacemacs/spacemacs-editing-visual/README.org"
Exporting "/tmp/docs/layers/+spacemacs/spacemacs-org/README.org" into "/tmp/spacetools/layers/+spacemacs/spacemacs-org/README.sdn"
Successfully exported "/tmp/docs/layers/+spacemacs/spacemacs-org/README.org"
Exporting "/tmp/docs/layers/+tools/prettier/README.org" into "/tmp/spacetools/layers/+tools/prettier/README.sdn"
Successfully exported "/tmp/docs/layers/+tools/prettier/README.org"
Exporting "/tmp/docs/layers/+spacemacs/spacemacs-completion/README.org" into "/tmp/spacetools/layers/+spacemacs/spacemacs-completion/README.sdn"
Successfully exported "/tmp/docs/layers/+spacemacs/spacemacs-completion/README.org"
Exporting "/tmp/docs/layers/+distributions/spacemacs-docker/deps-install/installers/javascript/README.org" into "/tmp/spacetools/layers/+distributions/spacemacs-docker/deps-install/installers/javascript/README.sdn"

=============== WARNING ===============
Missing "#+TAGS:" keyword. See "https://github.com/syl20bnr/spacemacs/blob/develop/core/templates/README.org.template"
=======================================

Successfully exported "/tmp/docs/layers/+distributions/spacemacs-docker/deps-install/installers/javascript/README.org"
Exporting "/tmp/docs/layers/+tools/command-log/README.org" into "/tmp/spacetools/layers/+tools/command-log/README.sdn"
Successfully exported "/tmp/docs/layers/+tools/command-log/README.org"
Exporting "/tmp/docs/layers/+fun/selectric/README.org" into "/tmp/spacetools/layers/+fun/selectric/README.sdn"
Successfully exported "/tmp/docs/layers/+fun/selectric/README.org"
Exporting "/tmp/docs/layers/+spacemacs/spacemacs-language/README.org" into "/tmp/spacetools/layers/+spacemacs/spacemacs-language/README.sdn"
Successfully exported "/tmp/docs/layers/+spacemacs/spacemacs-language/README.org"
Exporting "/tmp/docs/layers/+lang/gpu/README.org" into "/tmp/spacetools/layers/+lang/gpu/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/gpu/README.org"
Exporting "/tmp/docs/layers/+lang/pact/README.org" into "/tmp/spacetools/layers/+lang/pact/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/pact/README.org"
Exporting "/tmp/docs/layers/+fun/emoji/README.org" into "/tmp/spacetools/layers/+fun/emoji/README.sdn"
Successfully exported "/tmp/docs/layers/+fun/emoji/README.org"
Exporting "/tmp/docs/layers/+lang/csv/README.org" into "/tmp/spacetools/layers/+lang/csv/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/csv/README.org"
Exporting "/tmp/docs/layers/+lang/asciidoc/README.org" into "/tmp/spacetools/layers/+lang/asciidoc/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/asciidoc/README.org"
Exporting "/tmp/docs/layers/+lang/coffeescript/README.org" into "/tmp/spacetools/layers/+lang/coffeescript/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/coffeescript/README.org"
Exporting "/tmp/docs/layers/+misc/copy-as-format/README.org" into "/tmp/spacetools/layers/+misc/copy-as-format/README.sdn"
Successfully exported "/tmp/docs/layers/+misc/copy-as-format/README.org"
Exporting "/tmp/docs/layers/+pair-programming/floobits/README.org" into "/tmp/spacetools/layers/+pair-programming/floobits/README.sdn"
Successfully exported "/tmp/docs/layers/+pair-programming/floobits/README.org"
Exporting "/tmp/docs/layers/+misc/multiple-cursors/README.org" into "/tmp/spacetools/layers/+misc/multiple-cursors/README.sdn"
Successfully exported "/tmp/docs/layers/+misc/multiple-cursors/README.org"
Exporting "/tmp/docs/layers/+distributions/spacemacs-docker/deps-install/README.org" into "/tmp/spacetools/layers/+distributions/spacemacs-docker/deps-install/README.sdn"

=============== WARNING ===============
Missing "#+TAGS:" keyword. See "https://github.com/syl20bnr/spacemacs/blob/develop/core/templates/README.org.template"
=======================================

Successfully exported "/tmp/docs/layers/+distributions/spacemacs-docker/deps-install/README.org"
Exporting "/tmp/docs/layers/+web-services/evernote/README.org" into "/tmp/spacetools/layers/+web-services/evernote/README.sdn"
Successfully exported "/tmp/docs/layers/+web-services/evernote/README.org"
Exporting "/tmp/docs/layers/+lang/asm/README.org" into "/tmp/spacetools/layers/+lang/asm/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/asm/README.org"
Exporting "/tmp/docs/layers/+vim/vinegar/README.org" into "/tmp/spacetools/layers/+vim/vinegar/README.sdn"
Successfully exported "/tmp/docs/layers/+vim/vinegar/README.org"
Exporting "/tmp/docs/layers/+lang/json/README.org" into "/tmp/spacetools/layers/+lang/json/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/json/README.org"
Exporting "/tmp/docs/layers/+lang/racket/README.org" into "/tmp/spacetools/layers/+lang/racket/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/racket/README.org"
Exporting "/tmp/docs/layers/+tools/geolocation/README.org" into "/tmp/spacetools/layers/+tools/geolocation/README.sdn"
Successfully exported "/tmp/docs/layers/+tools/geolocation/README.org"
Exporting "/tmp/docs/layers/+lang/purescript/README.org" into "/tmp/spacetools/layers/+lang/purescript/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/purescript/README.org"
Exporting "/tmp/docs/layers/+lang/scala/README.org" into "/tmp/spacetools/layers/+lang/scala/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/scala/README.org"
Exporting "/tmp/docs/layers/+tools/ranger/README.org" into "/tmp/spacetools/layers/+tools/ranger/README.sdn"
Successfully exported "/tmp/docs/layers/+tools/ranger/README.org"
Exporting "/tmp/docs/layers/+frameworks/react/README.org" into "/tmp/spacetools/layers/+frameworks/react/README.sdn"
Successfully exported "/tmp/docs/layers/+frameworks/react/README.org"
Exporting "/tmp/docs/layers/+lang/ipython-notebook/README.org" into "/tmp/spacetools/layers/+lang/ipython-notebook/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/ipython-notebook/README.org"
Exporting "/tmp/docs/layers/+intl/chinese/README.org" into "/tmp/spacetools/layers/+intl/chinese/README.sdn"
Successfully exported "/tmp/docs/layers/+intl/chinese/README.org"
Exporting "/tmp/docs/layers/+lang/ruby/README.org" into "/tmp/spacetools/layers/+lang/ruby/README.sdn"
Successfully exported "/tmp/docs/layers/+lang/ruby/README.org"
Exporting "/tmp/docs/layers/+intl/keyboard-layout/README.org" into "/tmp/spacetools/layers/+intl/keyboard-layout/README.sdn"
Successfully exported "/tmp/docs/layers/+intl/keyboard-layout/README.org"
Exporting "/tmp/docs/.cache/quelpa/melpa/CONTRIBUTING.org" into "/tmp/spacetools/.cache/quelpa/melpa/CONTRIBUTING.sdn"

!!!!!!!!!!!!!!!! ERROR !!!!!!!!!!!!!!!!
current-buffer: CONTRIBUTING.org<2>
current-file: /tmp/docs/.cache/quelpa/melpa/CONTRIBUTING.org
error: Linked file "/tmp/docs/.cache/quelpa/melpa/README.md#recipe-format" not readable
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

From first view seems like Docker image needs attending.
</details>

**Comment by JAremko** [comment]
> From the first brief view seems like Docker image needs attending.

@Anton-Latukha   I usually run image against clean fork instead of my .emacs.d dir. But I'll add `.cache` to the ignore list.

**Comment by JAremko** [comment]
@Anton-Latukha  pushed new image.

**Comment by Anton-Latukha** [comment]
@JAremko Ok.
Yes, since it was a basic text/doc-based work I made it directly on the live distribution. & fork has old branch of **this* branch.

**Comment by Anton-Latukha** [comment]
@JAremko 
I also think it is a good moment to allow and start adding short descriptions near layers. Some layers that are a superset, like `major-modes` - obviously need attached description in the main list, to at least be discoverable.

**Comment by JAremko** [comment]
@Anton-Latukha  Descriptions from the README.org files are copied into the corresponding LAYERS.org entries. You can add additional information into them (descriptions)

**Comment by Anton-Latukha** [comment]
The situation I observe swung my decision.
Included the `Markup languages` into the `Domain-specific languages` category.

They are really belong there by computer science nomenclature. 
And since there are also `jsonnet`, `semantic-web`, `yang`, `dhall` they are blurring the line between simple markup languages, their templating, and programming, this languages demanded strict type hierarchy (place in DSL and Markup categories simultaneously).

**Comment by Anton-Latukha** [comment]
I would also include `Frameworks` as `Programming languages/Frameworks`.

**Comment by Anton-Latukha** [comment]
@JAremko 
If something my current config block looks:
```edn
  {"programming" ["util"
                  {"dsl" ["lisp"
                          "markup"
                          "script"
                          ]}
                  {"general" ["imperative"
                              {"multi-paradigm" ["js"
                                                 ]}
                              "pure"
                              ]}
                  ]}
```

**Comment by JAremko** [comment]
> All layers: lang: programming: general: multi-paradigm languages went into the black hole, and this I observe in all such cases, DSLs also. What would you say, how would we solve this situation. (my current structure config is posted below)

@Anton-Latukha   `[{"dsl" ["lisp"  "markup" "script"]} "dsl"]`

**Comment by Anton-Latukha** [comment]
Yes, the thing is - it lists there the whole subtree with subsets also.
I added a `root` tag.
We also can use Org italic `/*/` (aka _*_) to be more intuitive. In my exp - it works for Org files.

This is current state.

The thing I noticed - since `root` tag pops-up several times in listing - Org result does not distinguish between:
```org
  - [[#domain-specific-dsls][Domain-specific (DSLs)]]
    - [[#root][root]]
...
    - [[#multi-paradigm][Multi-paradigm]]
      - [[#root-1][root]]
```

And links all TOC `root` links to `[[#root][root]]`.
So I am think: or to solve it, or to supply unique tags.

**Comment by JAremko** [comment]
@Anton-Latukha  TOC links seems to work properly.

**Comment by Anton-Latukha** [comment]
Ok. About links - was an empty message, I should thought more. Talked about org-links inside Emacs & it doesn't metter in what we do. Since web links work properly.

About Org view, since I do not know how it would be parsed in HTML site form - I'd would increase TOC to `:TOC_5_gh:` to show layers in `multi-paradigm/root` and `multi-paradigm/js`.

**Comment by Anton-Latukha** [comment]
@JAremko 
So far so good. I am satisfied with the result. This chunk seems done.

**Comment by JAremko** [comment]
@Anton-Latukha I added layer sorting and now each README.org file can appear only once so `root` tags no longer needed. Also READMEs that haven't been matched by the query tree will be displayed at the bottom of the LAYERS file.

**Comment by JAremko** [comment]
@Anton-Latukha  Also this all needs a `git squash` into 1-3 commits.

**Comment by Anton-Latukha** [comment]
Checked TOC 4 -> 5 influenceю It is minimal, across all Spacemacs doc files in total added several (~5-7) headers.

**Comment by JAremko** [comment]
@Anton-Latukha  Are you sure that you used latest image ?  This seems like an old error https://circleci.com/gh/syl20bnr/spacemacs/32805?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link :thinking:

**Comment by JAremko** [comment]
@Anton-Latukha  I'll fix this https://circleci.com/gh/syl20bnr/spacemacs/32811?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link  and we can merge :thinking:

**Comment by Anton-Latukha** [comment]
Conf form:
```edn
  "pairing"
  "reader"
  {"programming" [{"dsl" ["lisp"
                          "markup"
                          "script"
                          ]}
                  "dsl"
                  {"general" ["imperative"
                              {"multi-paradigm" ["js"
                                                 ]}
                              "multi-paradigm"
                              "pure"
                              ]}
                  "general"
                  "framework"
                  "util"
                  ]}
  "versioning"
  {"spacemacs" ["distribution"
                "util"
                ]}
  "tag"
  "theme"
```
It is obvious that in this manner "Multiply identical path IDs ... can happen if two headlines have child headlines with too similar names."

**Comment by JAremko** [comment]
@Anton-Latukha  Yeah I'll fix it by making the tool merge same headlines.

**Comment by JAremko** [comment]
@Anton-Latukha  Should be fixed now (with the latest image)  Hopefully last update :stuck_out_tongue: 

Also don't forget to mention you https://circleci.com/gh/syl20bnr/spacemacs/32826?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link :wink:

**Comment by JAremko** [comment]
@Anton-Latukha  I cherry picked it. You can delete the PR branch. :+1:

**Comment by JAremko** [comment]
@Anton-Latukha  Btw there still a lot of work: https://github.com/syl20bnr/spacemacs/blob/develop/layers/LAYERS.org#readmeorg-files-that-need-proper-tags  those are layers that haven't been matched by the query tree.

**Pull request closed without merging**